Documentation/design-docs/dllmap.md

```diff
@@ -0,0 +1,246 @@
+	hmod = LocalLoadLibraryHelper(currLibNameVariation, dllImportSearchPathFlag, pErrorTracker)
+hmod = LocalLoadLibraryHelper(pModule->GetPath(), loadWithAlteredPathFlags | dllImportSearchPathFlag, pErrorTracker)
+```
+`LoadLibraryModuleViaHost`  already contains a callback (`AssemblyLoadContext` exposes `LoadUnmanagedDll()` API to load a dll but it can be used for a `CustomAssemblyLoadContext` only, not the default one).  `LoadLibraryViaCallback`  will do  a callback, but only for non-system assemblies. `System.Private.CoreLib` and other ‘system’ assemblies will not receive callbacks.
+
+### Entrypoint mapping
+Once hmod gets resolved and returned via `LoadLibraryModule`, an entrypoint must be find:
+
+Currently `NDirectGetEntryPoint` takes two arguments: `pMD` and hmod. When `NDirectGetEntryPoint`  gets called, `pMD `points to a target dll and hmod is correlated with a target dll too.
+The entrypoint name mapping will be done at the beginning of `NDirectGetEntryPoint`. A new method ` IntPtr  GetMappedEntrypoint(sourceEntrypointName)`  will be a callback and will return a mapping for an entrypoint if it exists. `pMD` will get updated to point to a target entrypoint. The rest of `NDirectGetEntryPoint()`  flow will remain the same.
+```c++
+HINSTANCE hmod = LoadLibraryModule( pMD, &errorTracker );
+if ( hmod )
+{
+	LPVOID pvTarget = GetEntrypointViaCallback(pMD, hmod); // this is the only introduced step
+	if (!pvTarget)
+	{
+		LPVOID pvTarget = NDirectGetEntryPoint(pMD, hmod);
+	}
+	…
+}
+```
+
+In consequence, after `LoadLibraryModule()`  and `GetEntrypointViaCallback() / NDirectGetEntryPoint()`  execution, `pMD` will get updated: `pMD->SetNDirectTarget(pvTarget)` with a correctly mapped `pvTarget`.
+
+### Callbacks
+
+As explained above, runtime will rise two dll specific events on each load attempt:
+* 1st callback - when loading a non-system library
+* 2nd callback - when finding an entrypoint
+
+Events will be defined in `AssemblyLoadContext` in `Private.CoreLib`. Default handlers that subscribe dll load events will implement the mono-based dllmap logic. They will take string as argument and return IntPtr of target libraries and entrypoints based on the parsed XML configuration file.
+
+Handlers implementation will stay in `AssemblyLoadContext`. Load library resolver will cache all the dll mapping results that got resolved (as key-value: IntPtr-hmod pairs). Thanks to that, the same library won't get loaded multiple times. User’s code will be able to subscribe to events and implement any loading behavior. That will give a user full flexibility when using dllmap and won’t limit defining the mapping to only xml-based style. Callbacks will be executed for non-system assemblies only. `System.Private.CoreLib` and other system assemblies will not receive callbacks, because that could cause problems like an infinite dll-mapping loop.
```

Comment:
> Callbacks will be executed for non-system assemblies only.

What will be the mechanism to define non-system assemblies?

Reply:
That’s tricky. I thought there was an easy way to determine this, e.g. some kind of check that we could execute on runtime. But it looks like there isn’t. 
A possible workaround is to enable callbacks for all assemblies and distinguish between “system” and “non-system” assemblies in the default handler. We could assume that when an assembly name starts with “System.*” then there is a system assembly, otherwise, it’s not. Relying on names is not very convenient, but I don’t have a better idea. Any suggestions? @luqunl 
Another possibility is to enable mapping for all assemblies, but it may cause problems.

Reply:
In the assembly loader we have this concept.  Take a look at https://github.com/dotnet/coreclr/blob/13e7c4368da664a8b50228b1a5ef01a660fbb2dd/src/vm/assembly.hpp#L103.  Since we are going to need similar information to get the assemblyloadcontext I am hopeful we will have access to the module manifest.

Reply:
This concept is not sufficient for this. For example, a lot of file I/O APIs - that one is very likely going to use in the callback - are implemented outside CoreLib.

Reply:
> For example, a lot of file I/O APIs - that one is very likely going to use in the callback

That actually raises a pretty interesting question: how is the user going to know what APIs are safe to call from these callbacks without causing infinite recursions *on all runtimes that support this*. I assume we're going to make sure whatever XML and file IO we need for the dllmap compatibility plugin works - but if someone comes up with an idea to do the resolution based on talking to a web service through WCF - I assume that won't be safe to do from the callback.

Reply:
The alternative way to solve this is to change the shape of the API: Change the API shape to register delegate per assembly instead of having a global event. Then the rule is simple: Make sure that the resolver won't call any PInvokes in the assembly you have registered it for.

Reply:
We discussed an alternative which is that these NativeDll resolve events are not reentrant.  If reentrancy is detected we will resolve the p/invoke via the usual manner and not call the callbacks.  This has the pro that it avoids people having to be 'careful' but it may not be easily discoverable on why p/invoke remapping does not occur 'sometimes'.

@jkotas if we explored an assembly level resolution, do you have a thought on what the callback would extend from?  System.Reflection.Assembly?  I had considered this and it frankly makes some cases a lot easier (for instance library providers could have large sole control of resolution).

Reply:
> if we explored an assembly level resolution, do you have a thought on what the callback would extend from? 

It can be a new type with a bunch of static methods like what is proposed in https://github.com/dotnet/coreclr/issues/19112 . It can have both the power-user methods that use callbacks and the convenience methods that just let you register the mapping.

Reply:
How about an assembly-level (or module-level) attribute people can place on their libraries that specifies a type that provides resolution services for the given assembly/module?

People can say something along the lines of this in the C# files:

```csharp
[assembly: NativeDllResolver(typeof(DllMapResolver))]
```

One could make it even more elegant once C# brings support for generic attributes:

```csharp
// NativeDllResolverAttribute can be constrained to implement an interface
// so that we don't have to worry about handling the case of resolver being bogus
[assembly: NativeDllResolver<DllMapResolver>]
```

Reply:
Based on the variety of scenarios on how people would like to leverage this feature, I feel an assembly level attribute could be one of a few ways to add a callback.  Though I feel we need to explore the other options as well.

> It can be a new type with a bunch of static methods like what is proposed in #19112

In this deisgn, I feel the type would need to be part of CoreApp to enable the right layering of the runtime calling back.  Are you suggesting that this type would then have the XML parsing, etc?

Reply:
> Are you suggesting that this type would then have the XML parsing, etc?

The xml parsing would be still in separate optional package. It would use these methods to register itself for the assemblies it wants to deal with.

Reply:
It may be useful to have the actual user stories written down and validate the design against them. Here are the two that I think are the most interesting (there can be more):

- Scenario 1: I have a managed library with custom native library probing logic. The only way I can make this work today is via marshaled delegates and a lot of boilerplate code. I want to be able to use a regular DllImport to reduce amount of boilerplate code multiple times and get better performance.
The real world example for this scenario is our own System.Drawing.Common: https://github.com/dotnet/corefx/blob/master/src/System.Drawing.Common/src/System/Drawing/GdiplusNative.cs

- Scenario 2: I have a large project that is targeting Mono and depends on dllmap heavily. I want to be able to do initial migratation of this project to .NET Core with minimal changes (some changes are ok).
The real world example for this scenario is https://github.com/mono/monodevelop on Mac (Visual Studio for Mac is built out of this codebase). It has several dozen dllmap files, All of them use just dllmap, none of them remap entrypoints. It suggests that the entrypoint remapping is a corner case feature and it may be ok to just not worry about it for this.

Reply:
> if we explored an assembly level resolution, do you have a thought on what the callback would extend from?

it can also add another extract parameter to current callback.

// managedAssemblyName is the assembly name which use PInvoke
// nativeLibraryName is the file name specified in DllImport
IntPtr  LoadNativeDllViaDllMap(string managedAssemblyName, string nativeLibraryName)